Framework/Algorithms/src/PDCalibration.cpp

```diff
@@ -544,22 +542,29 @@ void PDCalibration::exec() {
   const auto windowsInDSpacing =
       dSpacingWindows(m_peaksInDspacing, peakWindowMaxInDSpacing);
 
+  // get spectrum info to check workspace index correpsonds to a valid spectrum
+  const auto &spectrumInfo = m_uncalibratedWS->spectrumInfo();
+
   // Scan the table containing the fit parameters for every peak, retrieve the
   // parameters for peaks that were successfully fitting, then use this info
   // to obtain difc, difa, and tzero for each pixel
   // cppcheck-suppress syntaxError
-  PRAGMA_OMP(parallel for schedule(dynamic, 1) )
+  PRAGMA_OMP(parallel for schedule(dynamic, 1))
   for (int wkspIndex = 0; wkspIndex < NUMHIST; ++wkspIndex) {
     PARALLEL_START_INTERUPT_REGION
-    if (isEvent && uncalibratedEWS->getSpectrum(wkspIndex).empty()) {
+    if ((isEvent && uncalibratedEWS->getSpectrum(wkspIndex).empty()) ||
+        !spectrumInfo.hasDetectors(wkspIndex) ||
+        spectrumInfo.isMonitor(wkspIndex)) {
       prog.report();
       continue;
     }
 
     // object to hold the information about the peak positions, detid, and wksp
     // index
     PDCalibration::FittedPeaks peaks(m_uncalibratedWS, wkspIndex);
-    auto toTof = getDSpacingToTof(peaks.detid);
+    auto toTof = getDSpacingToTof(
+        *peaks.detid
+             .begin()); // doesn't matter which one - all have same difc etc.
```

Comment:
Is it true that all the detectors associated with a peak with all have the same diffractometer constants? Seems like it's more likely they'll all be different?

Reply:
In principle they have different difc as they are in different positions, but by virtue of the fact you are summing over them by specifying the grouping there can only be one set of constant refined (this was the solution approved by WISH people a while back). happy to discuss.

Reply:
Thinking about this a bit more... 
It seems like the spectrum may correspond to an entire bank with quite a long list of detectors. So feels like you probably should be taking the average difa, difc, tzero because the differences might be non-trivial across an entire bank?
There's some code in AlignDetectors in a class called ConversionFactors which does this already that you could copy.

Reply:
Yes I see what you mean, using an average will always give you a better estimate for the TOF peak centres.  In normal circumstances I don't think it would matter as the algorithm is intended for calibrating the whole instrument rather than focussed/grouped data with very different difc, however the Engineering diffraction workflow at ISIS is one instance where it would make a bigger difference.

Reply:
I think giving the grouped pixels all the same DIFC is the only reasonable thing to do. In principle, one can write an algorithm to do something more intelligent using the "common" DIFC as a starting point (smoothing between neighbouring groups, use cross-correlation on pixels within a group, etc).